src/ngraph/frontend/onnx_import/utils/reshape.cpp

```diff
@@ -52,40 +60,105 @@ namespace ngraph
 
                 return std::make_shared<ngraph::op::Reshape>(
-                    node,
-                    ngraph::AxisVector{input_order},
-                    ngraph::Shape{first_dim_size, last_dim_size});
+                    node, AxisVector{input_order}, Shape{first_dim_size, last_dim_size});
             }
-        } // namespace utils
 
-        std::shared_ptr<ngraph::Node> reorder_axes(const std::shared_ptr<ngraph::Node>& node,
-                                                   std::vector<size_t> axes_order = {})
-        {
-            ngraph::Shape out_shape = node->get_shape();
-            if (axes_order.empty())
+            AxisVector get_default_axis_vector(std::size_t data_shape_size, std::size_t start_value)
             {
-                axes_order.resize(out_shape.size());
-                std::iota(std::begin(axes_order), std::end(axes_order), 0);
+                AxisVector axis_vector(data_shape_size);
+                std::iota(std::begin(axis_vector), std::end(axis_vector), start_value);
+                return axis_vector;
             }
-            else
+
+            std::vector<std::size_t> infer_dimensions(const std::string& node_name,
+                                                      const std::vector<std::size_t>& input_shape,
+                                                      const std::vector<std::size_t>& output_shape)
             {
-                for (auto i = 0; i < axes_order.size(); ++i)
+                std::vector<std::size_t> inferred_dims{output_shape};
+
+                // If an output dimension is equal to zero its actual value is copied from the input
+                // shape argument.
+                for (std::size_t idx = 0; idx < inferred_dims.size(); ++idx)
                 {
-                    out_shape[i] = node->get_shape().at(axes_order.at(i));
+                    if (inferred_dims.at(idx) == 0)
+                    {
+                        if (idx < input_shape.size())
+                        {
+                            inferred_dims.at(idx) = input_shape.at(idx);
+                        }
+                        else
+                        {
+                            throw error::parameter::Value(
+                                "Reshape",
+                                node_name,
+                                "can not copy dimension from the input data shape since requested "
+                                "index is out of range.");
+                        }
+                    }
                 }
+
+                // Check whether there are dimensions equal to -1 in output_shape. There may be at most
+                // one such case. Its value is then inferred from the size of the tensor and the
+                // remaining dimensions.
+                auto neg_value_it =
+                    std::find(std::begin(inferred_dims), std::end(inferred_dims), -1);
+                auto sec_neg_value_it =
```

Comment:
You don't need to `find` again if `net_value_i` is equal `std::end(neg_value_it)`.

Reply:
```
auto it = std::find(std::begin(inferred_dims), std::end(inferred_dims), -1);
if (it != std::end(inferred_dims)
{
    if (std::find(std::next(it), std::end(inferred_dims), -1) != std::end(inferred_dims)))
    {
        throw error::parameter::Value{...};
    }
    ...
}
return inferred_dims;